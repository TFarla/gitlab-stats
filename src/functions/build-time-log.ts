import Commit from "../graphql/types/commit";

export interface Task {
  sha: string,
  message: string,
  issue: number,
  totalTimeInMinutes: number
}

export interface Log {
  day: Date,
  commits: Commit[],
  tasks: Task[]
}

type LogMap = Map<string, Log>;
const issueRegex = /#(?<issue>\d+)\+(?<time>\w+)/g;
const timeRegex = /((?<day>\d+)d|(?<hour>\d+)h|(?<minute>\d+)m)/g;

export const buildTimeLog = (commits: Commit[]) =>
  commits.reduce<LogMap>((acc, commit) => {
    const isoDate = toIsoDate(commit.committed_date);
    const dateString = isoDate.toISOString();
    const tasks = toTasks(commit);
    if (acc.has(dateString)) {
      const entry = acc.get(dateString);
      if (entry) {
        entry.commits.push(commit);
        entry.tasks = [...entry.tasks, ...tasks];
        return acc;
      }
    }

    return acc.set(dateString, {
      day: isoDate,
      commits: [commit],
      tasks
    });
  }, new Map());

function toIsoDate(isoDateString: string) {
  const date = new Date(isoDateString);
  return new Date(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate());
}

const toTasks = (commit: Commit) =>
  Array.from(commit.message.matchAll(issueRegex)).reduce<Task[]>((acc, match) => {
    if (match.groups) {
      acc.push({
        sha: commit.id,
        issue: parseInt(match.groups['issue']),
        message: commit.message,
        totalTimeInMinutes: humanReadableTimeToMinutes(match.groups['time'])
      })
    }
    return acc;
  }, []);


export const humanReadableTimeToMinutes = (time: string) =>
  Array.from(time.matchAll(timeRegex)).reduce<number>((minutes, match) => {
    if (match.groups) {
      const { day, hour, minute } = match.groups;
      const dayInMinutes = parseInt(day || '0') * 1440;
      const hourInMinutes = parseInt(hour || '0') * 60;
      return minutes + parseInt(minute || '0') + dayInMinutes + hourInMinutes;
    }

    return minutes;
  }, 0);
