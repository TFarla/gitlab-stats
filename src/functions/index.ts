export * from './build-time-log';
