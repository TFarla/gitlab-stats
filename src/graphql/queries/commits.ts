import { useQuery } from "@apollo/client";
import gql from "graphql-tag";
import Commit from "../types/commit";
import { Issue } from "../types/issue";

export const commitsQuery = gql`
  query commits($url: String!, $projectId: Int!, $apiKey: String!) {
    commits(url: $url, projectId: $projectId, apiKey: $apiKey)
    @rest(type: "Commit", path: "{args.url}/api/v4/projects/{args.projectId}/repository/commits/?private_token={args.apiKey}&all=true&per_page=100000") {
      id
      title
      message
      created_at
      committed_date
    }

    issues(url:$url, projectId: $projectId, apiKey: $apiKey)
    @rest(type: "Issue", path: "{args.url}/api/v4/projects/{args.projectId}/issues/?private_token={args.apiKey}&per_page=100") {
      id
      iid
      project_id
      title
      description
      milestone {
        iid
        title
      }
    }
  }
`;

interface Variables {
  url: string,
  projectId: number,
  apiKey: string
}

export interface CommitsQueryResult {
  commits: Commit[],
  issues: Issue[]
}

export const useCommits = (variables: Variables) => useQuery<CommitsQueryResult, Variables>(commitsQuery, {variables});

export default commitsQuery;
