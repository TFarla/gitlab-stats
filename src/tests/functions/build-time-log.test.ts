import { buildTimeLog, humanReadableTimeToMinutes } from "../../functions";
import commits from './commits.json';

describe('build-time-log', () => {
  it('should be empty on no commits', () => {
    const log = buildTimeLog([]);
    expect(Array.from(log)).toHaveLength(0);
  })

  it('should group commits', () => {
    expect(buildTimeLog(commits)).toMatchSnapshot();
  })

  it.each`
    time                | minutes
    ${'1h30m'}          | ${90}
    ${'2h'}             | ${120}
    ${'1d'}             | ${1440}
    ${'2h1h1h'}         | ${240}
    ${'nonsense #2+3h'} | ${180}
    `('should convert to time notation', ({ time, minutes }) => {
    expect(humanReadableTimeToMinutes(time)).toEqual(minutes);
  })
})
