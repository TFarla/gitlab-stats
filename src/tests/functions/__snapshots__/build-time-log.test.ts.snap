// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-time-log should group commits 1`] = `
Map {
  "2020-12-03T23:00:00.000Z" => Object {
    "commits": Array [
      Object {
        "author_email": "email@example.com",
        "author_name": "Thomas Farla",
        "authored_date": "2020-12-04T07:32:10.000+01:00",
        "committed_date": "2020-12-04T07:32:10.000+01:00",
        "committer_email": "email@example.com",
        "committer_name": "Thomas Farla",
        "created_at": "2020-12-04T07:32:10.000+01:00",
        "id": "6a672d339456554ae491f93c342b5c2e0f41c631",
        "message": "Setup project #1+1h
",
        "parent_ids": Array [
          "1a95648880a2b7b7710d02dcc570b4fe76b6836e",
        ],
        "short_id": "6a672d33",
        "title": "Setup project #1+1h",
        "web_url": "https://gitlab.com/TFarla/gitlab-stats/-/commit/6a672d339456554ae491f93c342b5c2e0f41c631",
      },
      Object {
        "author_email": "email@example.com",
        "author_name": "Thomas Farla",
        "authored_date": "2020-12-04T08:08:31.000+01:00",
        "committed_date": "2020-12-04T08:08:31.000+01:00",
        "committer_email": "email@example.com",
        "committer_name": "Thomas Farla",
        "created_at": "2020-12-04T08:08:31.000+01:00",
        "id": "d7b8832f4b8af542991069c2f2e3df52b104a072",
        "message": "Add whiteline to readme #2+1h
",
        "parent_ids": Array [
          "6a672d339456554ae491f93c342b5c2e0f41c631",
        ],
        "short_id": "d7b8832f",
        "title": "Add whiteline to readme #2+1h",
        "web_url": "https://gitlab.com/TFarla/gitlab-stats/-/commit/d7b8832f4b8af542991069c2f2e3df52b104a072",
      },
      Object {
        "author_email": "email@example.com",
        "author_name": "Thomas Farla",
        "authored_date": "2020-12-04T08:15:30.000+01:00",
        "committed_date": "2020-12-04T08:15:30.000+01:00",
        "committer_email": "email@example.com",
        "committer_name": "Thomas Farla",
        "created_at": "2020-12-04T08:15:30.000+01:00",
        "id": "a1ab41eb6a23ecbdcac28b943ddac99fd9037fce",
        "message": "Remove newline #2+1h
",
        "parent_ids": Array [
          "d7b8832f4b8af542991069c2f2e3df52b104a072",
        ],
        "short_id": "a1ab41eb",
        "title": "Remove newline #2+1h",
        "web_url": "https://gitlab.com/TFarla/gitlab-stats/-/commit/a1ab41eb6a23ecbdcac28b943ddac99fd9037fce",
      },
      Object {
        "author_email": "email@example.com",
        "author_name": "Thomas Farla",
        "authored_date": "2020-12-04T08:16:02.000+01:00",
        "committed_date": "2020-12-04T08:16:02.000+01:00",
        "committer_email": "email@example.com",
        "committer_name": "Thomas Farla",
        "created_at": "2020-12-04T08:16:02.000+01:00",
        "id": "98a53de677f1a23186dac0b979564659b121218e",
        "message": "Add newline #2+1h
",
        "parent_ids": Array [
          "a1ab41eb6a23ecbdcac28b943ddac99fd9037fce",
        ],
        "short_id": "98a53de6",
        "title": "Add newline #2+1h",
        "web_url": "https://gitlab.com/TFarla/gitlab-stats/-/commit/98a53de677f1a23186dac0b979564659b121218e",
      },
      Object {
        "author_email": "email@example.com",
        "author_name": "Thomas Farla",
        "authored_date": "2020-12-04T14:55:00.000+01:00",
        "committed_date": "2020-12-04T14:55:00.000+01:00",
        "committer_email": "email@example.com",
        "committer_name": "Thomas Farla",
        "created_at": "2020-12-04T14:55:00.000+01:00",
        "id": "bf45d20b4a06ac415187fdf7c91d33fd5a714e58",
        "message": "Fetch commits and issues from gitlab #2+4h
",
        "parent_ids": Array [
          "6a672d339456554ae491f93c342b5c2e0f41c631",
        ],
        "short_id": "bf45d20b",
        "title": "Fetch commits and issues from gitlab #2+4h",
        "web_url": "https://gitlab.com/TFarla/gitlab-stats/-/commit/bf45d20b4a06ac415187fdf7c91d33fd5a714e58",
      },
      Object {
        "author_email": "email@example.com",
        "author_name": "Thomas Farla",
        "authored_date": "2020-12-04T15:45:06.000+01:00",
        "committed_date": "2020-12-04T15:45:06.000+01:00",
        "committer_email": "email@example.com",
        "committer_name": "Thomas Farla",
        "created_at": "2020-12-04T15:45:06.000+01:00",
        "id": "22edd3fced215e8b4a17eb996c0f9ee7c7f48c05",
        "message": "Add commit heatmap #2+1h
",
        "parent_ids": Array [
          "bf45d20b4a06ac415187fdf7c91d33fd5a714e58",
        ],
        "short_id": "22edd3fc",
        "title": "Add commit heatmap #2+1h",
        "web_url": "https://gitlab.com/TFarla/gitlab-stats/-/commit/22edd3fced215e8b4a17eb996c0f9ee7c7f48c05",
      },
      Object {
        "author_email": "email@example.com",
        "author_name": "Thomas Farla",
        "authored_date": "2020-12-04T17:06:06.000+01:00",
        "committed_date": "2020-12-04T17:06:06.000+01:00",
        "committer_email": "email@example.com",
        "committer_name": "Thomas Farla",
        "created_at": "2020-12-04T17:06:06.000+01:00",
        "id": "4d6ab3be33590e33a28fb542f30aa99bc419587d",
        "message": "Add project picker #3+1h
",
        "parent_ids": Array [
          "22edd3fced215e8b4a17eb996c0f9ee7c7f48c05",
        ],
        "short_id": "4d6ab3be",
        "title": "Add project picker #3+1h",
        "web_url": "https://gitlab.com/TFarla/gitlab-stats/-/commit/4d6ab3be33590e33a28fb542f30aa99bc419587d",
      },
    ],
    "day": 2020-12-03T23:00:00.000Z,
    "tasks": Array [
      Object {
        "issue": 1,
        "message": "Setup project #1+1h
",
        "sha": "6a672d339456554ae491f93c342b5c2e0f41c631",
        "totalTimeInMinutes": 60,
      },
      Object {
        "issue": 2,
        "message": "Add whiteline to readme #2+1h
",
        "sha": "d7b8832f4b8af542991069c2f2e3df52b104a072",
        "totalTimeInMinutes": 60,
      },
      Object {
        "issue": 2,
        "message": "Remove newline #2+1h
",
        "sha": "a1ab41eb6a23ecbdcac28b943ddac99fd9037fce",
        "totalTimeInMinutes": 60,
      },
      Object {
        "issue": 2,
        "message": "Add newline #2+1h
",
        "sha": "98a53de677f1a23186dac0b979564659b121218e",
        "totalTimeInMinutes": 60,
      },
      Object {
        "issue": 2,
        "message": "Fetch commits and issues from gitlab #2+4h
",
        "sha": "bf45d20b4a06ac415187fdf7c91d33fd5a714e58",
        "totalTimeInMinutes": 240,
      },
      Object {
        "issue": 2,
        "message": "Add commit heatmap #2+1h
",
        "sha": "22edd3fced215e8b4a17eb996c0f9ee7c7f48c05",
        "totalTimeInMinutes": 60,
      },
      Object {
        "issue": 3,
        "message": "Add project picker #3+1h
",
        "sha": "4d6ab3be33590e33a28fb542f30aa99bc419587d",
        "totalTimeInMinutes": 60,
      },
    ],
  },
}
`;
